### PR TITLE
83409 Increase E2E testing for form 10-10d

### PR DESCRIPTION
**Pull request by michaelclement**
## Summary

This PR adds several new E2E tests. Tests were run in loops to verify they are not flaky. They cover the following courses:

Test 1
- The user is an applicant
- The sponsor is alive
- The applicant is the sponsor's step child
- The applicant has Medicare A/B
- The applicant does not have Medicare D or OHI

Test 2
- The user is an applicant
- The sponsor is deceased
- The applicant is the sponsor's biological child
- The applicant has OHI
- The applicant does not have any form of Medicare

Test 3
- The user is a 3rd party certifying on behalf of an applicant
- The sponsor is deceased
- the applicant was married to the sponsor but divorced

Test 4
- This is a maximal test that includes a 3rd party certifier, a living sponsor, and three applicants with most conditional paths included

Test 5
- The user is the sponsor
- Two applicants are included
- Applicant 1 is the sponsor's spouse, has Medicare A/B/D, and OHI
- Applicant 2 is the sponsor's biological child and does not have Medicare or OHI

Summary
- Team: IVC CHAMPVA
- Flipper: NA

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/83409

## Testing done

- E2E, unit 

## What areas of the site does it impact?

IVC form 10-10d

### Quality Assurance & Testing

- [X] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [X] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [X] Browser console contains no warnings or errors.
- [X] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

NA

### CODE REVIEW

File: src/applications/ivc-champva/10-10D/config/form.js [hunk 1]
```diff
@@ -1039,7 +1039,7 @@ const formConfig = {
         },
         // Applicant separated from sponsor before sponsor's death
         page18f6: {
-          path: 'applicant-information/:index/married-separated-dates',
+          path: 'applicant-information/married-separated-dates/:index',
```
Comment (michaelclement) [inline src/applications/ivc-champva/10-10D/config/form.js:1042]: Updated this URL to match pattern used elsewhere in form (index on end)

File: src/applications/ivc-champva/10-10D/tests/e2e/fixtures/data/app-sa-cs-medab.json [hunk 2]
```diff
@@ -0,0 +1,92 @@
+{
+  "description": "Is applicant, sponsor alive, child (step), has Medicare A/B",
```
Comment (michaelclement) [inline src/applications/ivc-champva/10-10D/tests/e2e/fixtures/data/app-sa-cs-medab.json:2]: This property summarizes the key points of this data-set. It is also reflected in the filename (abbreviated)

Comment (michaelclement) [reply src/applications/ivc-champva/10-10D/tests/e2e/fixtures/data/app-sa-cs-medab.json:2]: This is done for the other datasets as well.

File: src/applications/ivc-champva/10-7959C/tests/10-7959C.cypress.spec.js [hunk 3]
```diff
@@ -9,16 +9,12 @@ import manifest from '../manifest.json';
...
 
 // Put all page objects into an object where pagename maps to page data
 // E.g., {page1: {path: '/blah'}}
-const ALL_PAGES = {};
-Object.values(formConfig.chapters).forEach(ch =>
-  Object.keys(ch.pages).forEach(p => {
-    ALL_PAGES[p] = ch.pages[p];
-  }),
-);
+const ALL_PAGES = getAllPages(formConfig);
```
Comment (michaelclement) [inline src/applications/ivc-champva/10-7959C/tests/10-7959C.cypress.spec.js:17]: Moved this up to be a shared function between all IVC forms

File: src/applications/ivc-champva/10-7959C/tests/10-7959C.cypress.spec.js [hunk 4]
```diff
@@ -128,9 +124,13 @@ const testConfig = createTestConfig(
               // format of dates (goes from YYYY-MM-DD to MM-DD-YYYY).
               // TODO: Address discrepancy at some point.
               expect(data[k]?.length).to.equal(req.body[k]?.length);
+            } else {
+              expect(data[k] === req.body[k]);
```
Comment (michaelclement) [inline src/applications/ivc-champva/10-7959C/tests/10-7959C.cypress.spec.js:128]: Base case where we just check that the data input into the form matches what we get on the other side.

File: src/applications/ivc-champva/shared/components/applicantLists/ApplicantRelationshipPage.jsx [hunk 5]
```diff
@@ -306,6 +306,7 @@ export default function ApplicantRelationshipPage({
                     customOtherDescription ||
                     `Since ${relativePossessive} relationship with the ${personTitle} was not listed, please describe it here`
                   }
+                  name="other-relationship-description"
```
Comment (michaelclement) [inline src/applications/ivc-champva/shared/components/applicantLists/ApplicantRelationshipPage.jsx:309]: This component was missing an identifier, so just added this for clarity

File: src/applications/ivc-champva/shared/tests/helpers.js [hunk 6]
```diff
@@ -41,7 +41,14 @@ export const fillAddressWebComponentPattern = (fieldName, addressObject) => {
...
   fillTextWebComponent(`${fieldName}_street3`, addressObject.street3);
-  selectDropdownWebComponent(`${fieldName}_state`, addressObject.state);
+  // List loop fields sometimes fail on this because the state <select> renders as a text input
+  // TODO: look into that bug. For now, set the test to check which field type we have
+  cy.get('body').then(body => {
+    if (body.find(`va-select[name="root_${fieldName}_state"]`).length > 0)
+      selectDropdownWebComponent(`${fieldName}_state`, addressObject.state);
+    if (body.find(`va-text-input[name="root_${fieldName}_state"]`).length > 0)
+      fillTextWebComponent(`${fieldName}_state`, addressObject.state);
+  });
```
Comment (michaelclement) [inline src/applications/ivc-champva/shared/tests/helpers.js:51]: Workaround for a bug in the current form-lib implementation of addresses in the list loop. Will need to track this issue elsewhere.

**Review by reiting — APPROVED** [review]

**Review by cloudmagic80 — APPROVED** [review]

**Review by balexandr — APPROVED** [review]

**Review by harshilmoradia — APPROVED** [review]

**Pull request merged**